src/plugins/embeddable/public/lib/state_transfer/embeddable_state_transfer.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { cloneDeep } from 'lodash';
+import { ScopedHistory, ApplicationStart } from '../../../../../core/public';
+import {
+  EmbeddableOriginatingAppState,
+  isEmbeddableOriginatingAppState,
+  EmbeddablePackageState,
+  isEmbeddablePackageState,
+} from './types';
+
+export class EmbeddableStateTransfer {
+  constructor(private navigateToApp: ApplicationStart['navigateToApp']) {}
+
+  public getIncomingOriginatingApp(
+    history: ScopedHistory,
+    removeAfterFetch: boolean = false
+  ): EmbeddableOriginatingAppState | undefined {
+    return this.getIncomingState<EmbeddableOriginatingAppState>(
+      history,
+      isEmbeddableOriginatingAppState,
+      removeAfterFetch
+    );
+  }
+
+  public getIncomingEmbeddablePackage(
```

Comment:
It's not a big problem and definitely optional, but a lot of these methods take a scoped history object to act on - this means that in the future if `stateTransfer` is used somewhere in a consumer, it's always necessary to pass down two things - the embeddable contract and the history object, then "marry" them in the call when necessary. 

Instead, there could be just a factory function `getStateTransferController(history: ScopedHistory)` on the contract of the embeddable plugin - this returns an `EmbeddableStateTransfer` instance which already has the `history` bound, so you can just call `getIncomingEmbeddablePackage(removeAfterFetch: boolean = false)`. That makes it possible to create the state transfer controller up in the tree of as consuming plugin (e.g. directly in the mount callback of the registered app) and only pass the stateful controller object down to the actual consumers.

E.g. this case https://github.com/elastic/kibana/pull/67064/files#diff-7b2c1602740d8df9744799bc2f1e5a54R74 would get a bit simpler. Again, not relevant right now, but it seems more in line with issues like https://github.com/elastic/kibana/issues/67999 which try to limit the amount of references which need to be passed around.

Reply:
Thanks for the suggestion Joe! I implemented this, even though it didn't manifest in many changes to this PR due to the fact that all places incoming stateTransfer is used already contain easy access to scopedHistory.

For the future, especially if this strongly typed wrapper makes its way into core, (Stay tuned for an issue on this) this sort of pattern will be really helpful.